### ISSUE TITLE
match with tuple erroneously reports reportMatchNotExhaustive

### ISSUE THREAD

**Issue by davetapley** [microsoft/pyright#7472]
**Describe the bug**

`match` with tuple erroneously reports `reportMatchNotExhaustive` 

**Code or Screenshots**
```py
from random import choice


class Foo:
    ...


class Bar:
    ...


def foobar() -> Foo | Bar:
    return choice([Foo(), Bar()])


a = foobar()
b = foobar()

match (a, b):
    case (Foo(), Foo()):
        ...
    case (Foo(), Bar()):
        ...
    case (Bar(), Foo()):
        ...
    case (Bar(), Bar()):
        ...
```

```
error: Cases within match statement do not exhaustively handle all values
    Unhandled type: "tuple[Foo | Bar, Foo | Bar]"
    If exhaustive handling is not intended, add "case _: pass" (reportMatchNotExhaustive)
1 error, 0 warnings, 0 informations 
```

**VS Code extension or command-line**

Both, `1.1.353`

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug. See [this issue](https://github.com/microsoft/pyright/issues/6437) for details.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.374.
